Add post nonlinear solver signal

gassmoeller opened the pull request:
This is mostly here for testing purposes and for showing this to @jdannberg and @tjhei for the time stepping discussion. I will update the PR for the other solver schemes and see if any tests change (they should not).

tjhei commented on source/simulator/solver_schemes.cc:
```diff
@@ -672,12 +672,22 @@ namespace aspect
   void Simulator<dim>::solve_single_advection_single_stokes ()
   {
+    // Setup a nonlinear solver control that only allows a single iteration
+    SolverControl nonlinear_solver(1,
```
Using a SolverControl is a good idea!

tjhei commented on include/aspect/simulator_signals.h:
```diff
@@ -250,6 +250,16 @@ namespace aspect
+    * about the final state (failure/success), number of 
+    * iterations and history of residuals.
+    */
+    boost::signals2::signal<void (const SimulatorAccess<dim> &,
```
We normally don't pass SimulatorAccess to the signals because there are other ways to get access to it. Maybe remove?

gassmoeller commented:
I added the solver control and signal to all solver schemes, which also allowed me to unify the nonlinear iteration loop in the schemes. I think the only thing left to do (tomorrow) is to add a test where the nonlinear solver fails and check that the state and residual are reported correctly.

tjhei commented on source/simulator/solver_schemes.cc:
```diff
@@ -694,25 +700,30 @@ namespace aspect
+    SolverControl nonlinear_solver_control(max_nonlinear_iterations,
+                                           parameters.nonlinear_tolerance,
+                                           true,
+                                           true);
```
I would remove the two bool args, the defaults are fine (here and in all other places).

tjhei commented on source/simulator/solver_schemes.cc:
```diff
@@ -908,24 +952,25 @@ namespace aspect
-    newton_handler->parameters.preconditioner_stabilization = starting_preconditioner_stabilization;
-    newton_handler->parameters.velocity_block_stabilization = starting_velocity_block_stabilization;
+  // Reset the Newton stabilization at the end of the timestep.
+  newton_handler->parameters.preconditioner_stabilization = starting_preconditioner_stabilization;
```
what happened with the indentation here?

bangerth replied:
I think you lost a closing scope brace?

bangerth commented on include/aspect/simulator_signals.h:
```diff
@@ -250,6 +250,14 @@ namespace aspect
                                   const SolverControl &solver_control)> post_advection_solver;
 
+    /**
+    * A signal that is fired when the nonlinear solver scheme is done.
```
Might want to align the asterisks.

bangerth commented on include/aspect/simulator_signals.h:
```diff
@@ -250,6 +250,14 @@ namespace aspect
+    * about the final state (failure/success), number of
+    * iterations and history of residuals of the nonlinear solver.
+    */
+    boost::signals2::signal<void (const SolverControl &)> post_nonlinear_solver;
```
I think you want to document that this signal is also called for the linear schemes that do not actually iterate, and in that case will simply record a single iteration with a success outcome.

bangerth commented on source/simulator/solver_schemes.cc:
```diff
@@ -694,25 +700,30 @@ namespace aspect
+    double relative_residual = std::numeric_limits<double>::max();
+    for (nonlinear_iteration = 0;
+         nonlinear_solver_control.check(nonlinear_iteration, relative_residual) == SolverControl::iterate;
+         ++nonlinear_iteration)
```
I think this would be easier to read as a `while` loop:
```
  nonlinear_iteration = 0;
  while (control.check(...) == ...)
  {
    ...
    ++nonlinear_iteration;
  }
```

bangerth replied:
Or maybe even easier, via the `do ... while(...)` construct.

bangerth commented on source/simulator/solver_schemes.cc:
```diff
@@ -749,7 +772,15 @@ namespace aspect
+    double relative_residual = std::numeric_limits<double>::max();
+    for (nonlinear_iteration = 0;
+         nonlinear_solver_control.check(nonlinear_iteration, relative_residual) == SolverControl::iterate;
+         ++nonlinear_iteration)
```
same here

bangerth commented on source/simulator/solver_schemes.cc:
```diff
@@ -822,26 +849,29 @@ namespace aspect
+    double relative_residual = std::numeric_limits<double>::max();
+    for (nonlinear_iteration = 0;
+         nonlinear_solver_control.check(nonlinear_iteration, relative_residual) == SolverControl::iterate;
+         ++nonlinear_iteration)
```
and here

bangerth commented on source/simulator/solver_schemes.cc:
```diff
@@ -882,13 +912,27 @@ namespace aspect
+    double relative_residual = std::numeric_limits<double>::max();
+    for (nonlinear_iteration = 0;
+         nonlinear_solver_control.check(nonlinear_iteration, relative_residual) == SolverControl::iterate;
+         ++nonlinear_iteration)
```
and here

bangerth commented on source/simulator/solver_schemes.cc:
```diff
@@ -966,13 +1011,28 @@ namespace aspect
+    double relative_residual = std::numeric_limits<double>::max();
+    for (nonlinear_iteration = 0;
+         nonlinear_solver_control.check(nonlinear_iteration, relative_residual) == SolverControl::iterate;
+         ++nonlinear_iteration)
```
and here

gassmoeller commented:
I think I addressed all comments. To be consistent with our current test results I had to create one `while` loop and all other solver schemes as `do ... while` loops, because some tests rely on the fact that no iteration is done if `max_nonlinear_iterations == 0`.

tjhei commented on source/simulator/solver_schemes.cc:
```diff
@@ -882,13 +912,23 @@ namespace aspect
+    double relative_residual = std::numeric_limits<double>::max();
+    nonlinear_iteration = 0;
+
+    while (nonlinear_solver_control.check(nonlinear_iteration, relative_residual) == SolverControl::iterate)
```
why is this one a while loop while you use do ... while in all other places?

gassmoeller replied:
Because a) it was that way before as well (it was a for loop before), and b) we have some tests that assume this is a while loop (they set 0 max nonlinear iterations and expect to never go into the loop) and other tests that require all other schemes to not be a while loop (they require that the check is only executed after the nonlinear solve, and the number of nonlinear residuals recorded is 1 after 1 solve). I could make it work with consistent do .. while loops if I added a special condition that skipped the loop if max_nonlinear_iterations == 0, would you prefer that instead to keep the loops consistent?

tjhei replied:
I would think setting the nonlinear iterations to 0 is a bug and should fail. What do you think?

bangerth replied:
Is that the total number of nonlinear steps (Picard + Newton)? That seems wrong -- what is the solution vector supposed  to be if that happens?

bangerth commented on source/simulator/solver_schemes.cc:
```diff
@@ -678,6 +678,12 @@ namespace aspect
+    // Setup a nonlinear solver control that only allows a single iteration
+    SolverControl nonlinear_solver_control(1,1.0);
+    // Announce that we did a single iteration, and assume we have converged
+    nonlinear_solver_control.check(0,0.0);
```
I think you want
```suggestion
    nonlinear_solver_control.check(1, 0.0);
```
to indicate that *one* iteration was actually done.

bangerth commented on source/simulator/solver_schemes.cc:
```diff
@@ -1059,6 +1115,12 @@ namespace aspect
+    // Setup a nonlinear solver control that only allows a single iteration
+    SolverControl nonlinear_solver_control(1,1.0);
+    // Announce that we did a single iteration, and assume we have converged
+    nonlinear_solver_control.check(0,0.0);
```
same here

bangerth commented on source/simulator/solver_schemes.cc:
```diff
@@ -1068,6 +1130,12 @@ namespace aspect
+    // Setup a nonlinear solver control that only allows a single iteration
+    SolverControl nonlinear_solver_control(1,1.0);
+    // Announce that we did a single iteration, and assume we have converged
+    nonlinear_solver_control.check(0,0.0);
```
and here

tjhei commented:
I will give @bangerth a chance to comment before merging.

bangerth commented:
Looks good, thanks!
